File: Difficulty: Medium/Segregate even and odd nodes in a Linked List/segregate-even-and-odd-nodes-in-a-linked-list.cpp
// User function template for C++

/*
struct Node
{
    int data;
    struct Node* next;

    Node(int x){
        data = x;
        next = NULL;
    }
};
*/
class Solution {
  public:
    Node* divide(Node* head) {
        // code here
         if (head == NULL || head->next == NULL) {
            return head;
        }
        
        Node *evenStart = NULL, *evenEnd = NULL;
        Node *oddStart = NULL, *oddEnd = NULL;
        Node *current = head;
        
        while (current != NULL) {
            if (current->data % 2 == 0) {
                if (evenStart == NULL) {
                    evenStart = current;
                    evenEnd = evenStart;
                } else {
                    evenEnd->next = current;
                    evenEnd = evenEnd->next;
                }
            } else {
                if (oddStart == NULL) {
                    oddStart = current;
                    oddEnd = oddStart;
                } else {
                    oddEnd->next = current;
                    oddEnd = oddEnd->next;
                }
            }
            current = current->next;
        }
        
        if (evenStart == NULL) {
            return oddStart;
        }
        
        evenEnd->next = oddStart;
        if (oddEnd != NULL) {
            oddEnd->next = NULL;
        }
        
        return evenStart;
    }
};

File: Difficulty: Medium/Segregate even and odd nodes in a Linked List/segregate-even-and-odd-nodes-in-a-linked-list.py
# User function Template for Python3

# Following is the structure of node 
# class node:
#     def __init__(self):  
#         self.data = None
#         self.next = None

class Solution:
    def divide(self, head):
        # code here
        es=node()
        os=node()
        
        ee= es
        oe= os
        
        temp = head
        while temp is not None:
            if temp.data%2==0:
                ee.next = temp
                ee= ee.next
            else:
                oe.next = temp
                oe = oe.next
            
            temp = temp.next
        oe.next = None
        ee.next = os.next
        return es.next